staging/vhost-device-console/src/backend.rs

```diff
@@ -0,0 +1,240 @@
+            //
+            // There isn't much value in complicating code here to return an error from the
+            // threads, and so the code uses unwrap() instead. The panic on a thread won't cause
+            // trouble to other threads/guests or the main() function and should be safe for the
+            // daemon.
+
+            let controller = ConsoleController::new(console_path.clone())
+                .map_err(Error::CouldNotCreateConsoleController)?;
+            let arc_controller = Arc::new(RwLock::new(controller));
+            let vu_console_backend = Arc::new(RwLock::new(
+                VhostUserConsoleBackend::new(arc_controller)
+                    .map_err(Error::CouldNotCreateBackend)?,
+            ));
+
+            let mut daemon = VhostUserDaemon::new(
+                String::from("vhost-device-console-backend"),
+                vu_console_backend.clone(),
+                GuestMemoryAtomic::new(GuestMemoryMmap::new()),
+            )
+            .map_err(Error::CouldNotCreateDaemon)?;
+
+            /* Start the read thread -- need to handle it after termination */
+            let vring_workers = daemon.get_epoll_handlers();
+            vu_console_backend
+                .read()
+                .unwrap()
+                .set_vring_worker(&vring_workers[0]);
+            let read_handle = VhostUserConsoleBackend::start_console_thread(&vu_console_backend);
+
+            let listener = Listener::new(socket.clone(), true).unwrap();
+            daemon.start(listener).unwrap();
+
+            match daemon.wait() {
+                Ok(()) => {
+                    info!("Stopping cleanly.");
+                }
+                Err(vhost_user_backend::Error::HandleRequest(
+                    vhost_user::Error::PartialMessage | vhost_user::Error::Disconnected,
+                )) => {
+                    info!("vhost-user connection closed with partial message. If the VM is shutting down, this is expected behavior; otherwise, it might be a bug.");
+                }
+                Err(e) => {
+                    warn!("Error running daemon: {:?}", e);
+                }
+            }
+
+            // No matter the result, we need to shut down the worker thread.
+            vu_console_backend
+                .read()
+                .unwrap()
+                .exit_event
+                .write(1)
+                .unwrap();
+
+            // Kill console input thread
+            vu_console_backend
+                .read()
+                .unwrap()
+                .kill_console_thread();
+
+            // Wait for read thread to exit
+            match read_handle.join() {
+                Ok(_) => info!("The read thread returned successfully"),
+                Err(e) => warn!("The read thread returned the error: {:?}", e),
+            }
+        });
+
+        handles.push(handle);
+    }
+
+    for handle in handles {
+        handle.join().map_err(|_| Error::FailedJoiningThreads)??;
+    }
```

Comment:
We discovered this leads to problems when a thread dies but the code here is blocked on a thread handle on a smaller index in `handles`.

See a better approach here using channels to signal thread shutdown: https://github.com/rust-vmm/vhost-device/blame/5a2bd27037ef1efdcfdffa733e8c34c25204d6e5/vhost-device-input/src/main.rs#L125-L161